[RISCV][Clang] Added code to generate clang error if immediate operan…

adeel10x opened the pull request:
…d of the following builtins is out of range:

-  __builtin_riscv_cv_simd_extract_h
-  __builtin_riscv_cv_simd_extract_b
-  __builtin_riscv_cv_simd_extractu_h
-  __builtin_riscv_cv_simd_extractu_b
-  __builtin_riscv_cv_simd_insert_h
-  __builtin_riscv_cv_simd_insert_b

adeel10x committed:
[RISCV][Clang] Added code to generate clang error if immediate operand of the following builtins is out of range:
-  __builtin_riscv_cv_simd_extract_h
-  __builtin_riscv_cv_simd_extract_b
-  __builtin_riscv_cv_simd_extractu_h
-  __builtin_riscv_cv_simd_extractu_b
-  __builtin_riscv_cv_simd_insert_h
-  __builtin_riscv_cv_simd_insert_b

adeel10x committed:
clang formatted

PaoloS02 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -4461,6 +4462,23 @@ bool Sema::CheckRISCVBuiltinFunctionCall(const TargetInfo &TI,
+  case RISCVCOREV::BI__builtin_riscv_cv_simd_extract_b:
+  case RISCVCOREV::BI__builtin_riscv_cv_simd_extractu_h:
+  case RISCVCOREV::BI__builtin_riscv_cv_simd_extractu_b:
+    return SemaBuiltinConstantArgRange(TheCall, 1, 0, 63);
```
I don't think we should use 63 as range limit. These instructions are supposed to be used with values [0,1,2,3] for *_b and [0,1] *_h to select the bytes or half words. I'd make that clear straight away in the front-end, otherwise we may need to truncate or crash or silently mis-select the byte/half word later. So I'd say:
for extract_h and extractu_h:
```
  return SemaBuiltinConstantArgRange(TheCall, 1, 0, 1);
```
while for extract_b and extractu_b:
```
  return SemaBuiltinConstantArgRange(TheCall, 1, 0, 3);
```

adeel10x replied:
Thanks for the feedback. I'll make the changes accordingly.

adeel10x replied:
I have fixed the immediate operand ranges. I'll edit the tests now.

adeel10x replied:
I think I clang-formatted the file too. Is it fine? Or should I revert the changes?

PaoloS02 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -4461,6 +4462,23 @@ bool Sema::CheckRISCVBuiltinFunctionCall(const TargetInfo &TI,
+    return SemaBuiltinConstantArgRange(TheCall, 1, 0, 63);
+  case RISCVCOREV::BI__builtin_riscv_cv_simd_insert_h:
+  case RISCVCOREV::BI__builtin_riscv_cv_simd_insert_b:
+    return SemaBuiltinConstantArgRange(TheCall, 2, 0, 63);
```
Same fix as above for the range

adeel10x replied:
I have fixed the ranges accordingly. I just noticed that I have run clang-format on the files. Is it fine or should I reverse it?

PaoloS02 commented:
I've added a couple of comments but otherwise it looks good. Could you also check whether there are Sema tests for this kind of validations? We could do with a test for these

PaoloS02 commented:
> I've added a couple of comments but otherwise it looks good. Could you also check whether there are Sema tests for this kind of validations? We could do with a test for these

I see there are not many Sema tests so probably the best way to test the errors is to add cases to `CodeGen/RISCV/corev-intrinsics/simd.c`  where we check that the error message is generated.

adeel10x commented:
> > I've added a couple of comments but otherwise it looks good. Could you also check whether there are Sema tests for this kind of validations? We could do with a test for these
> 
> I see there are not many Sema tests so probably the best way to test the errors is to add cases to `CodeGen/RISCV/corev-intrinsics/simd.c` where we check that the error message is generated.

ok sure.

adeel10x committed:
[RISCV][Clang]Fixed immediate operand range for the following intrinsics:
-  __builtin_riscv_cv_simd_extract_h
-  __builtin_riscv_cv_simd_extract_b
-  __builtin_riscv_cv_simd_extractu_h
-  __builtin_riscv_cv_simd_extractu_b
-  __builtin_riscv_cv_simd_insert_h
-  __builtin_riscv_cv_simd_insert_b

adeel10x commented:
> > I've added a couple of comments but otherwise it looks good. Could you also check whether there are Sema tests for this kind of validations? We could do with a test for these
> 
> I see there are not many Sema tests so probably the best way to test the errors is to add cases to `CodeGen/RISCV/corev-intrinsics/simd.c` where we check that the error message is generated.

Hi @Paolo,
I have one confusion. If I add failing cases to simd.c, wouldn't that cause the file to not compile causing all other cases to fail?
So, shouldn't I create a new file e.g: simd-error.c, and test for erroneous code there? Please let me know if I'm missing something here.
Thanks

PaoloS02 commented:
Legit question. You'll need a separate test like `clang/test/CodeGen/RISCV/rvv_errors.c`  to just check the ranges., so yes, a simd-errors.c test sounds like a good idea.

PaoloS02 commented:
On a separate note, I see that the diffs contain a lot of indentation changes. Has some formatting happened? Would it be possible to reload a patch with just your changes?

adeel10x commented:
> On a separate note, I see that the diffs contain a lot of indentation changes. Has some formatting happened? Would it be possible to reload a patch with just your changes?

Yeah, sure. I ran clang-format by mistake (force of habit). I'll revert it.

adeel10x committed:
Revert "clang formatted"

This reverts commit 1c372bceb0b4438492a3c95e72a31af23b261540.

adeel10x committed:
[RISCV][Clang] Added tests to check the immediate operand range for the following intrinsics:
    -  __builtin_riscv_cv_simd_extract_h
    -  __builtin_riscv_cv_simd_extract_b
    -  __builtin_riscv_cv_simd_extractu_h
    -  __builtin_riscv_cv_simd_extractu_b
    -  __builtin_riscv_cv_simd_insert_h
    -  __builtin_riscv_cv_simd_insert_b

adeel10x commented:
> Legit question. You'll need a separate test like `clang/test/CodeGen/RISCV/rvv_errors.c` to just check the ranges., so yes, a simd-errors.c test sounds like a good idea.

I have added the tests.

adeel10x commented:
Hi @PaoloS02 , 
There are some [tests](https://github.com/openhwgroup/corev-llvm-project/blob/4535d99b4e8dac33f3e7182bbb34d100b8c2f2b0/clang/test/CodeGen/RISCV/corev-intrinsics/simd.c#L2259C1-L2353C2) for **__builtin_riscv_cv_simd_extract*** in **simd.c**. These tests will fail after the recent changes as they are using values beyond the allowed range. Should I remove  __builtin_riscv_cv_simd_extract*  tests that are using a negative value for the immediate operand from simd.c, as I've already added such tests in **simd-errors.c.**

Also, I think I should adjust the value of the immediate operand for those __builtin_riscv_cv_simd_extract* tests in simd.c that are using positive immediate value but the value is out of valid range.

PaoloS02 commented:
Hi @adeel10x sorry for the delay. Yes to all the above. Since we have a separate test to cover the error cases so that we can see whether future changes break the check on the operand all we need in the simd.c test is normal use cases with appropriate operands to check that the selection of the intrinsics works. Thanks for this!

PaoloS02 commented:
The new tests look good!

adeel10x committed:
Modified tests for the following builtins:
__builtin_riscv_cv_simd_extract_h
__builtin_riscv_cv_simd_extract_b
__builtin_riscv_cv_simd_extractu_h
__builtin_riscv_cv_simd_extractu_b
